src/plugins/mavlink_ftp/mavlink_ftp_impl.h

```diff
@@ -0,0 +1,166 @@
+#pragma once
+
+#include <fstream>
+#include <mutex>
+#include <string>
+#include <dirent.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "mavlink_include.h"
+#include "plugins/mavlink_ftp/mavlink_ftp.h"
+#include "plugin_impl_base.h"
+
+namespace mavsdk {
+
+class MavlinkFTPImpl : public PluginImplBase {
+public:
+    /// @brief Command opcodes
+    enum Opcode : uint8_t {
+        kCmdNone,             ///< ignored, always acked
+        kCmdTerminateSession, ///< Terminates open Read session
+        kCmdResetSessions,    ///< Terminates all open Read sessions
+        kCmdListDirectory,    ///< List files in <path> from <offset>
+        kCmdOpenFileRO,       ///< Opens file at <path> for reading, returns <session>
+        kCmdReadFile,         ///< Reads <size> bytes from <offset> in <session>
+        kCmdCreateFile,       ///< Creates file at <path> for writing, returns <session>
+        kCmdWriteFile,        ///< Writes <size> bytes to <offset> in <session>
+        kCmdRemoveFile,       ///< Remove file at <path>
+        kCmdCreateDirectory,  ///< Creates directory at <path>
+        kCmdRemoveDirectory,  ///< Removes Directory at <path>, must be empty
+        kCmdOpenFileWO,       ///< Opens file at <path> for writing, returns <session>
+        kCmdTruncateFile,     ///< Truncate file at <path> to <offset> length
+        kCmdRename,           ///< Rename <path1> to <path2>
+        kCmdCalcFileCRC32,    ///< Calculate CRC32 for file at <path>
+        kCmdBurstReadFile,    ///< Burst download session file
+
+        kRspAck = 128,        ///< Ack response
+        kRspNak               ///< Nak response
+    };
+
+    MavlinkFTPImpl(System &system);
+    MavlinkFTPImpl(const MavlinkFTPImpl &) = delete;
+    const MavlinkFTPImpl &operator=(const MavlinkFTPImpl &) = delete;
+
+    ~MavlinkFTPImpl();
+
+    void init() override;
+    void deinit() override;
+
+    void enable() override;
+    void disable() override;
+
+    void send();
+
+    void download_async(std::string remote_file_path,
```

Comment:
Did you choose to copy the `std::string` on purpose?

Reply:
Changed - std::string s are passed by reference.